src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+  int bytesLeftToCompare = 0;
+  int offsetOfFirstByteToCompare = 0;
+
+  bytesAlreadyCompared = isU ? (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE * 2) : NUMBER_OF_NEEDLE_BYTES_TO_COMPARE;
+  offsetOfFirstByteToCompare
+      = isU ? ((NUMBER_OF_NEEDLE_BYTES_TO_COMPARE - 1) * 2) : (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE - 1);
+
+  bytesLeftToCompare = size - bytesAlreadyCompared;
+  assert((bytesLeftToCompare <= 8), "Too many bytes left to compare");
+
+  if (bytesLeftToCompare <= 0) {
+    return;
+  }
+
+  // At this point, there is at least one byte of the needle that needs to be
+  // compared to the haystack.
+
+  if (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE <= 2) {
+    switch (bytesLeftToCompare) {
+    case 1:
+    case 2:
+      __ movzwl(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    case 3:
+    case 4:
+      __ movl(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    case 5:
+      // Read one byte before start of needle, then mask it off
+      __ movq(needleVal, Address(needle, (offsetOfFirstByteToCompare - 2)));
+      __ shrq(needleVal, 0x8);
+      break;
+
+    case 6:
+      __ movq(needleVal, Address(needle, (offsetOfFirstByteToCompare - 1)));
+      break;
+
+    case 7:
+    case 8:
+      __ movq(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    default:
+      break;
+    }
+  }
+
+  if (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE >= 3) {
+    switch (bytesLeftToCompare) {
+    case 1:
+    case 2:
+      __ movl(needleVal, Address(needle, (offsetOfFirstByteToCompare - 2)));
+      break;
+
+    case 3:
+    case 4:
+      __ movl(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    case 5:
+    case 6:
+      __ movq(needleVal, Address(needle, (offsetOfFirstByteToCompare - 2)));
+      break;
+
+    case 7:
+    case 8:
+      __ movq(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    default:
+      break;
+    }
+  }
+}
+
+static void byte_compare_helper(int size, Label &L_noMatch, Label &L_matchFound, Register needle, Register needleVal,
+                                Register haystack, Register mask, Register foundIndex, Register tmp,
+                                StrIntrinsicNode::ArgEncoding ae, MacroAssembler *_masm) {
+  // Compare size bytes of needle to haystack
+  //
+  // At a minimum, the first and last bytes of needle already compare equal
+  // to the haystack, so there is no need to compare them again.
+
+  Label L_loopTop;
+
+  assert_different_registers(needle, needleVal, haystack, mask, foundIndex, tmp);
+
+  bool isLL = (ae == StrIntrinsicNode::LL);
+  bool isUL = (ae == StrIntrinsicNode::UL);
+  bool isUU = (ae == StrIntrinsicNode::UU);
+  bool isU = isUL || isUU;  // At least one is UTF-16
+
+  int bytesAlreadyCompared = 0;
+  int bytesLeftToCompare = 0;
+  int offsetOfFirstByteToCompare = 0;
+
+  Label temp;
+
+  if (isU) {
+    if ((size & 1) != 0) {
+      __ emit_int8(0xcc);
```

Comment:
Compile-time assert to ensure this code is never called instead?

Reply:
Done